### DISCUSSION TITLE
How do I adjust an existing form for two use cases?

### DISCUSSION THREAD

**Discussion by JanaS87** [neuefische/questions#697] [Web]
Hello everyone,

Currently, I am working on creating the possibility to make the information of an existing cat editable. For this, I have already changed the useState to useLocalStorageState() in the _app.js and built corresponding handler functions.

I have meanwhile thrown my original considerations overboard because they were simply too complex for me.
Currently, I have an UpdateCatForm, but I am considering whether it is necessary and am toying with the idea of putting everything into the already existing NewCatForm.

However, I do not know for both forms what I need to change so that everything works and is displayed as I have imagined, so I will probably delete the UpdateCatForm again and try to adapt the existing one (I have no idea what I´m doing =/ )

Do you have any ideas or can give me direct approaches on how I need to adjust everything? I lose track extremely quickly and with my concentration and focus problems, I am currently not getting anywhere.

This is my PR: [PR](https://github.com/JanaS87/capstone-project/pull/18)
It's not much code, actually its copy, delete, add code, delete again and so on... 

Thanks in advance for your help <3

**Comment by F-Kirchhoff** [comment]
Hey Jana,
der Trick besteht darin, die Props deines Forms geschickt zu verwenden. Du brauchst 2 Props:
- initialData - das was die Input felder am Anfang anzeigen sollen
- onSubmit callback - das was passieren soll, wenn der user auf den "submit" button drückt

Dann setzt du den "defaultValue" von jedem input Feld auf den jeweiligen Wert, der dir als prop reingereicht wird und rufst die von aussen als prop bereitgestellte Funktion "onSubmit" auf wenn der user auf den Button drückt. 

Jetzt kannst du ausserhalb deiner Component "Form" in den Props die Werte übergeben, die du an den jeweiligen Stellen brauchst und auch komplett andere "onSubmit" Funktionen verwenden. Ich habe dir das mal in einem kleinen 
Schaubild aufgezeichnet: 
![general-form](https://github.com/neuefische/questions/assets/59687330/f6cf54ae-ec86-445b-a599-ac912ffb971b)
Falls du ein Beispiel brauchst kannst du dir gerne das Handout zu "Backend update and delete" nochmal anschauen. 
Hoffe das hat geholfen.

**Reply by JanaS87** [reply]
Danke für die Antwort 🙂 
Zu dem Beispiel in unserem Handout, ich habe Probleme damit, das so umzusetzen, wie ich das benötige. Ich habe ja keine API und keine DB, ich versuche mir vorzustellen wie ich das umschreiben muss aber das macht mir Knoten ins Hirn 😿

**Reply by F-Kirchhoff** [reply]
irgendwo in deiner Form rufst du ja eine Funktion auf, in der du useLocalStorageState verwendest, oder? Diese Zeilen von Code musst du "von aussen" als eine separate Funktion über die Prop "onSubmit" reinreichen.

**Reply by JanaS87** [reply]
der useLocalStorage liegt in der _app. dort habe ich auch die functions geschrieben. Ich komme nicht weiter, weil ich die Übersicht in der Form verliere. Ich weiß nicht was ich davon nicht brauche, brauche oder umschreiben muss ohne, das sich die Fehler stapeln. Ich werde da einiges auslagern müssen (andere US) aber fürs erste ist die Logik dahinter wichtig und die geht mir komplett verloren irgendwie.

**Reply by F-Kirchhoff** [reply]
ah verstehe, dann würde ich an deiner stelle sämtliche Funktionalität aus deiner Form ausschneiden, sodass du nur noch das "html" hast. Dann kannst du dich schritt für schritt an die einzelnen Inputfelder machen und sie wieder von da aus aufbauen. Manchmal ist ein Neustart dafür das beste 😅

**Comment by ninaassmann** [comment]
So habe ich es gemacht:

ich habe mir in pages/pets (hier habe ich eine [slug].js für die Detailseite meiner Haustiere einen weiteren Ordner angelegt
also pages/pets/update und auch hier habe ich eine [slug].js angelegt
die beinhaltet meine Form Komponente (die Selbe wie die zum erstellen) und ich übergebe der Form als prop das pet, welches ich über useRouter mit dem slug bekomme

in der form befülle ich dann das defaultValue der einzelnen inputs mit den Inhalten des pet Objects, falls es eines gibt

Bsp.
defaultValue={pet && pet.petName}

wenn du einige Inputs nicht angezeigt haben möchtest könntest du sowas vielleicht machen
{!pet && <input>...}
bedeutet, dass er das input nur rendert, wenn kein pet vorhanden ist => !pet

alternativ könntest du mit dem diabled arbeiten
disabled={pet}
wenn pet vorhanden ist das input disabled und nicht bearbeitbar, ACHTUNG benötigt trotzdem das defaultValue, weil es sonst als leer weitergereicht wird
